overlord/fdestate/fdestate.go

```diff
@@ -0,0 +1,305 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package fdestate
+
+import (
+	"crypto"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/osutil/disks"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/secboot"
+)
+
+var (
+	disksDmCryptUUIDFromMountPoint = disks.DmCryptUUIDFromMountPoint
+	secbootGetPrimaryKeyHash       = secboot.GetPrimaryKeyHash
+	secbootVerifyPrimaryKeyHash    = secboot.VerifyPrimaryKeyHash
+)
+
+type Model struct {
+	SeriesValue    string             `json:"series"`
+	BrandIDValue   string             `json:"brand-id"`
+	ModelValue     string             `json:"model"`
+	ClassicValue   bool               `json:"classic"`
+	GradeValue     asserts.ModelGrade `json:"grade"`
+	SignKeyIDValue string             `json:"sign-key-id"`
+}
+
+func (m *Model) Series() string {
+	return m.SeriesValue
+}
+
+func (m *Model) BrandID() string {
+	return m.BrandIDValue
+}
+
+func (m *Model) Model() string {
+	return m.ModelValue
+}
+
+func (m *Model) Classic() bool {
+	return m.ClassicValue
+}
+
+func (m *Model) Grade() asserts.ModelGrade {
+	return m.GradeValue
+}
+
+func (m *Model) SignKeyID() string {
+	return m.SignKeyIDValue
+}
+
+func newModel(m secboot.ModelForSealing) Model {
+	return Model{
+		SeriesValue:    m.Series(),
+		BrandIDValue:   m.BrandID(),
+		ModelValue:     m.Model(),
+		ClassicValue:   m.Classic(),
+		GradeValue:     m.Grade(),
+		SignKeyIDValue: m.SignKeyID(),
+	}
+}
+
+var _ secboot.ModelForSealing = (*Model)(nil)
+
+type KeyroleSlotParameters struct {
+	Models         []Model         `json:"models,omitempty"`
+	BootModes      []string        `json:"boot-modes,omitempty"`
+	Tpm2PCRProfile json.RawMessage `json:"tpm2-pcr-profile,omitempty"`
+}
+
+type KeyslotRoleInfo struct {
+	// PrimaryKeyId is the ID for the primary key found in
+	// PrimaryKeys field of FdeState
+	PrimaryKeyId int `json:"primary-key-id"`
+	// Parameters is indexed by container role name
+	Parameters map[string]KeyroleSlotParameters `json:"params,omitempty"`
+	// Tpm2PCRPolicyRevocationCounter is the handle for the TPM
+	// policy revocation counter.  A value of 0 means it is not
+	// set.
+	Tpm2PCRPolicyRevocationCounter uint32 `json:"tpm2-pcr-policy-revocation-counter,omitempty"`
+}
+
+type hashAlg crypto.Hash
+
+func (h *hashAlg) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+	switch s {
+	case "sha256":
+		*h = hashAlg(crypto.SHA256)
+	default:
+		return fmt.Errorf("Unknown algorithm %s", s)
+	}
+
+	return nil
+}
+
+func (h hashAlg) MarshalJSON() ([]byte, error) {
+	switch crypto.Hash(h) {
+	case crypto.SHA256:
+		return json.Marshal("sha256")
+	default:
+		return nil, fmt.Errorf("Unknown algorithm %v", h)
+	}
+}
+
+type KeyDigest struct {
```

Comment:
This might be a type that should be handled in secboot. And we just use json.RawMessage for it.